include/tvm/runtime/container.h

```diff
@@ -0,0 +1,220 @@
+ *        and ElemType is handled by the memory allocator.
+ *
+ * @tparam ArrayType
+ * @tparam ElemType
+ */
+template <typename ArrayType, typename ElemType>
+class InplaceArrayBase {
+ public:
+  /**
+   * @brief Initialize the elements in the array.
+   */
+  void Init() {
+    CHECK_EQ(sizeof(ArrayType) % alignof(ElemType), 0);
+    for (size_t i = 0; i < Self()->size(); ++i) {
+      void* field_ptr = AddressOf(i);
+      new (field_ptr) ElemType();
+    }
+  }
+
+  /**
+   * @brief Initialize the elements in the array.
+   *
+   * @tparam Iterator Iterator type of the array.
+   * @param begin The begin iterator.
+   * @param end The end iterator.
+   */
+  template <typename Iterator>
+  void Init(Iterator begin, Iterator end) {
+    CHECK_EQ(sizeof(ArrayType) % alignof(ElemType), 0);
+    ArrayType* self = Self();
+    size_t num_elems = std::distance(begin, end);
+    if (num_elems != self->size()) {
+      LOG(FATAL)
+          << "Number of initializer values does not match number of elements\n";
+    }
+    auto it = begin;
+    for (size_t i = 0; i < num_elems; ++i) {
+      void* field_ptr = AddressOf(i);
+      new (field_ptr) ElemType(*it);
+      ++it;
+    }
+  }
+
+  /**
+   * @brief Initialize the elements in the array.
+   *
+   * @param init The initializer list of elements.
+   */
+  void Init(std::initializer_list<ElemType> init) {
+    CHECK_EQ(sizeof(ArrayType) % alignof(ElemType), 0);
+    Init(init.begin(), init.end());
+  }
+
+  /*!
+   * \brief Access element at index
+   * \param idx The index of the element.
+   * \return Reference to ElemType at the index.
+   */
+  ElemType& operator[](size_t idx) const {
+    size_t size = Self()->size();
+    if (idx > size) {
+      LOG(FATAL) << "Index " << idx << " out of bounds " << size << "\n";
+    }
+    return *(reinterpret_cast<ElemType*>(AddressOf(idx)));
+  }
+
+  /**
+   * @brief Destroy the Inplace Array Base object
+   */
+  virtual ~InplaceArrayBase() {
+    if (!IsPOD()) {
+      size_t size = Self()->size();
+      for (size_t i = 0; i < size; ++i) {
+        ElemType* fp = reinterpret_cast<ElemType*>(AddressOf(i));
+        fp->ElemType::~ElemType();
+      }
+    }
+  }
+
+ private:
+  /**
+   * @brief Check if the ElemType is Plain Old Data.
+   *
+   * @return If ElemType is POD.
+   */
+  inline bool IsPOD() const {
+    return std::is_standard_layout<ElemType>::value &&
+           std::is_trivial<ElemType>::value;
+  }
+
+  /**
+   * @brief Return the self object for the array.
+   *
+   * @return Pointer to ArrayType.
+   */
+  inline ArrayType* Self() const {
+    return static_cast<ArrayType*>(const_cast<InplaceArrayBase*>(this));
+  }
+
+  /**
+   * @brief Return the raw pointer to the element at idx.
+   *
+   * @param idx The index of the element.
+   * @return Raw pointer to the element.
+   */
+  void* AddressOf(size_t idx) const {
+    const size_t kDataStart = sizeof(ArrayType);
+    ArrayType* self = Self();
```

Comment:
You need to run the alignment static assertions somewhere to make sure the kDataStart gives you the right alignment